src/hb-serialize.hh

```diff
@@ -119,24 +119,41 @@ struct hb_serialize_context_t
 
   bool in_error () const { return !this->successful; }
 
+  void reset (void *start_, unsigned int size)
+  {
+    start = (char*) start_;
+    end = start + size;
+    reset ();
+    current = nullptr;
+  }
+
   void reset ()
   {
     this->successful = true;
     this->ran_out_of_room = false;
+    this->offset_overflow = false;
     this->head = this->start;
     this->tail = this->end;
     this->debug_depth = 0;
 
     fini ();
     this->packed.push (nullptr);
+    this->packed_map.init ();
   }
 
   bool check_success (bool success)
   { return this->successful && (success || (err_other_error (), false)); }
 
   template <typename T1, typename T2>
   bool check_equal (T1 &&v1, T2 &&v2)
-  { return check_success ((long long) v1 == (long long) v2); }
+  {
+    if ((long long) v1 != (long long) v2)
+    {
+      err_offset_overflow ();
+      return false;
+    }
+    return true;
+  }
```

Comment:
So this changes `check_equal()` to always result in an offset-overflow error. While we currently don't call this in any other instance, the assumption is not good.

Let's add a flag-enum for serialization error, which would be "ran-out-of-room" error, "other" error, and this new "offset overflow" error. And then all of `check_success()`, `check_equal()`, `check_assign()`, etc. take an optional error type.